semantic-reranker/src/main/java/org/opensearch/search/relevance/actionfilter/SearchActionFilter.java

```diff
@@ -0,0 +1,211 @@
+
+      @Override
+      public void onResponse(final Response response) {
+        final SearchResponse searchResponse = (SearchResponse) response;
+        final long totalHits = searchResponse.getHits().getTotalHits().value;
+        if (totalHits == 0) {
+          LOGGER.info("TotalHits = 0. Returning search response without semantic reranking: {}", searchResponse);
+          listener.onResponse(response);
+          return;
+        }
+
+        LOGGER.info("Starting semantic reranking for search response: {}", searchResponse);
+        // TODO: Do we need to write to a stream and parse from it?
+        try {
+          final BytesStreamOutput out = new BytesStreamOutput();
+          searchResponse.writeTo(out);
+
+          final StreamInput in = new NamedWriteableAwareStreamInput(out.bytes().streamInput(), namedWriteableRegistry);
+
+          LOGGER.info("Extracting search hits");
+          final SearchHits hits = new SearchHits(in);
+          final SearchHits newHits = doSemanticRerank(hits);
+
+          LOGGER.info("Extracting remaining response fields");
+          final InternalAggregations aggregations = in.readBoolean() ? InternalAggregations.readFrom(in) : null;
+          final Suggest suggest = in.readBoolean() ? new Suggest(in) : null;
+          final boolean timedOut = in.readBoolean();
+          final Boolean terminatedEarly = in.readOptionalBoolean();
+          final SearchProfileShardResults profileResults = in.readOptionalWriteable(SearchProfileShardResults::new);
+          final int numReducePhases = in.readVInt();
+
+          final SearchResponseSections internalResponse = new InternalSearchResponse(newHits, aggregations, suggest,
+              profileResults, timedOut, terminatedEarly, numReducePhases);
+
+          final int totalShards = in.readVInt();
+          final int successfulShards = in.readVInt();
+          final int shardSearchFailureSize = in.readVInt();
+          final ShardSearchFailure[] shardFailures;
+          if (shardSearchFailureSize == 0) {
+            shardFailures = ShardSearchFailure.EMPTY_ARRAY;
+          } else {
+            shardFailures = new ShardSearchFailure[shardSearchFailureSize];
+            for (int i = 0; i < shardFailures.length; i++) {
+              shardFailures[i] = readShardSearchFailure(in);
+            }
+          }
+
+          final SearchResponse.Clusters clusters = new SearchResponse.Clusters(in.readVInt(), in.readVInt(), in.readVInt());
+          final String scrollId = in.readOptionalString();
+          final int skippedShards = in.readVInt();
+
+          final long tookInMillis = (System.nanoTime() - startTime) / 1000000;
+
+          LOGGER.info("Creating new search response");
+          @SuppressWarnings("unchecked") final Response newResponse = (Response) new SearchResponse(internalResponse, scrollId, totalShards, successfulShards,
+              skippedShards, tookInMillis, shardFailures, clusters);
```

Comment:
Is there a way to only override the hits in the search response instead of creating a new search response? I feel there will be a lot of room for errors if we try to re-build the response with most of the same fields.